### PR TITLE
[WIP] Window iterator

### PR DESCRIPTION
**Pull request by kdmurray91**
Hi all,

This PR implements an iterator over `K`-length windows of a (initially) `Vector{T}`. This is required for (and discussed in) #103 

I intend to implement unit tests, and eventually extend this to work on things that aren't vectors.

Cheers,
K

### CODE REVIEW

**Comment by Ward9250** [comment]
Awesome, later today I might get time to have a play.

**Comment by kdmurray91** [comment]
The following should work

```julia
v = [1:30;]
for win in eachwindow(5, v)
    println(mean(win))
end
```

File: src/util/windows.jl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+end
+
+function eachwindow{T}(K::Integer, vec::Vector{T}, step::Integer=1)
+    @assert K >= 0 "K (window size) must be ≥ 0 in EachWindow"
+    @assert step >= 1 "step must be ≥ 1"
+
+    return EachWindowIterator{T}(vec, K, step)
+end
+
+function start{T}(it::EachWindowIterator{T})
```
Comment (Ward9250) [inline src/util/windows.jl]: Can we @inline these functions? I recall that inlining iterator methods increases performance by a not negligible margin.

Comment (kdmurray91) [inline src/util/windows.jl]: Oh yeah, I forgot about that. I originally had them inline, but I was on a train and julia-0.4 wasn't building properly so I reverted to 0.3 and had to remove the `@inline`. Restored in dd52090

Comment (kdmurray91) [inline src/util/windows.jl]: Doh, wrong branch. make that 40a1903

**Comment by bicycle1885** [comment]
I think there is room for generalization:

* We may be interested in a specific region in a sequence. So, adding start and stop positions to the iterator type and sliding within this range would be useful. Setting start position to 1 and stop position to the end of a sequence as default values would be reasonable.
* Sequence type can be generalized. The current type is limited to the `Vector{T}` type but we may want to iterate other type of sequences like `AminoAdidSequence` type. Parameterizing the sequence type is one way to do that. For example:

```julia
immutable EachWindowInterator{S}
    seq::S
    K::Int
    ...
```

**Comment by kdmurray91** [comment]
@bicycle1885 thanks for the comments!

> - We may be interested in a specific region in a sequence. So, adding start and stop positions to the iterator type and sliding within this range would be useful. Setting start position to 1 and stop position to the end of a sequence as default values would be reasonable.

Could this be done using `for w in eachwindow(5, seq[4:20]) ... end`? 

About your second point, I completely agree, I intend to generalise it to anything which implements `sub(x, i::UnitRange)`, which is what I currently use to generate windows. Which I think would mean that we need to implement `sub(seq, i:UnitRange)` for any sequence type we care about going through in windows, which I'd say would be all of them (including `SeqRecord` and subclasses). Does that sound right?

**Comment by Ward9250** [comment]
I would probalby do it something like:

`eachwindow(seq, range::Range, size::Int, step::Int)`

**Comment by bicycle1885** [comment]
> Could this be done using for w in eachwindow(5, seq[4:20]) ... end?

Exactly. Never mind my comment :sweat_smile: 

About the second point, we need careful consideration.
Given that `sub` creates a view object of an underlying object, I think we need to keep mutability of a sequence. That means a view of a mutable sequence is also mutable and mutating the view object also should mutate the original sequence.
This may cause problems because a view object may accidentally change an immutable sequence.

**Comment by Ward9250** [comment]
Will mutability be a problem for our seq sequences? Aren't they immutable by convention, but able to be set as mutable. So behaviour depends on that.

**Comment by Ward9250** [comment]
How are things progressing with this? I've been given a task that involves sliding windows, so if you don't have time to finish this, I can try and take it furthur.

**Comment by kdmurray91** [comment]
@Ward9250 I probably don't have much time for the next week or so, but regardless I'd love to see how someone more experienced with Julia improves upon my work. So go for it mate!

**Comment by Ward9250** [comment]
The related redux pull request #116, has been completed and merged.

**Pull request closed without merging**